src/commonMain/kotlin/dev/romainguy/kotlin/math/Quaternion.kt

```diff
@@ -0,0 +1,332 @@
+    }
+
+    operator fun get(
+            index1: QuaternionComponent,
+            index2: QuaternionComponent,
+            index3: QuaternionComponent): Float3 {
+        return Float3(get(index1), get(index2), get(index3))
+    }
+
+    operator fun get(
+            index1: QuaternionComponent,
+            index2: QuaternionComponent,
+            index3: QuaternionComponent,
+            index4: QuaternionComponent): Quaternion {
+        return Quaternion(get(index1), get(index2), get(index3), get(index4))
+    }
+
+    operator fun get(index: Int) = when (index) {
+        0 -> x
+        1 -> y
+        2 -> z
+        3 -> w
+        else -> throw IllegalArgumentException("index must be in 0..3")
+    }
+
+    operator fun get(index1: Int, index2: Int, index3: Int): Float3 {
+        return Float3(get(index1), get(index2), get(index3))
+    }
+
+    operator fun get(index1: Int, index2: Int, index3: Int, index4: Int): Quaternion {
+        return Quaternion(get(index1), get(index2), get(index3), get(index4))
+    }
+
+    inline operator fun invoke(index: Int) = get(index - 1)
+
+    operator fun set(index: Int, v: Float) = when (index) {
+        0 -> x = v
+        1 -> y = v
+        2 -> z = v
+        3 -> w = v
+        else -> throw IllegalArgumentException("index must be in 0..3")
+    }
+
+    operator fun set(index1: Int, index2: Int, v: Float) {
+        set(index1, v)
+        set(index2, v)
+    }
+
+    operator fun set(index1: Int, index2: Int, index3: Int, v: Float) {
+        set(index1, v)
+        set(index2, v)
+        set(index3, v)
+    }
+
+    operator fun set(index1: Int, index2: Int, index3: Int, index4: Int, v: Float) {
+        set(index1, v)
+        set(index2, v)
+        set(index3, v)
+        set(index4, v)
+    }
+
+    operator fun set(index: QuaternionComponent, v: Float) = when (index) {
+        QuaternionComponent.X -> x = v
+        QuaternionComponent.Y -> y = v
+        QuaternionComponent.Z -> z = v
+        QuaternionComponent.W -> w = v
+    }
+
+    operator fun set(index1: QuaternionComponent, index2: QuaternionComponent, v: Float) {
+        set(index1, v)
+        set(index2, v)
+    }
+
+    operator fun set(
+            index1: QuaternionComponent, index2: QuaternionComponent, index3: QuaternionComponent, v: Float) {
+        set(index1, v)
+        set(index2, v)
+        set(index3, v)
+    }
+
+    operator fun set(
+            index1: QuaternionComponent, index2: QuaternionComponent,
+            index3: QuaternionComponent, index4: QuaternionComponent, v: Float) {
+        set(index1, v)
+        set(index2, v)
+        set(index3, v)
+        set(index4, v)
+    }
+
+    operator fun unaryMinus() = Quaternion(-x, -y, -z, -w)
+
+    inline operator fun plus(v: Float) = Quaternion(x + v, y + v, z + v, w + v)
+    inline operator fun minus(v: Float) = Quaternion(x - v, y - v, z - v, w - v)
+    inline operator fun times(v: Float) = Quaternion(x * v, y * v, z * v, w * v)
```

Comment:
Are `times` and `div` valid quaternion operations as written here?

Reply:
As describe here: [Multiplying a Quaternion by a Scalar](https://www.3dgep.com/understanding-quaternions/#multiplying-a-quaternion-by-a-scalar)
If I well understood, the multiplacation must be applied to the scalar and to the vector so the functions must be valid
But I don't know if the div make sense.